### ISSUE TITLE
`WriteBufferFromS3` does one or two extra copies.

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#34346]
First copy happens here:

```
void WriteBufferFromS3::nextImpl()
{
    ...
    temporary_buffer->write(working_buffer.begin(), offset());
```

We already have data in the buffer. But we are copying it into another temporary buffer for no reason.

Second copy happens inside `std::basic_ostream` if I remember it correctly.
The basic rule for good C++ code is to avoid `iostream` from the standard library.
Also keep in mind, that by default, `std::basic_ostream` is not exception safe (if it performs allocation and it failed, the data will be silently truncated).

**Comment by KochetovNicolai** [comment]
It's in ASW sdk:
https://github.com/ClickHouse-Extras/aws-sdk-cpp/blob/266d7682c3d07d41618c41a3521811e1219c1bc7/aws-cpp-sdk-core/include/aws/core/AmazonStreamingWebServiceRequest.h#L44

**Comment by kitaisreal** [comment]
@KochetovNicolai we can provide own std buffer, or std stream implementation that will avoid additional copying.

CC: @alexey-milovidov

**Comment by alexey-milovidov** [comment]
Yes, plan A is to implement our streambuf and plan B is to bypass AWS SDK library in this place.
